astropy/io/fits/hdu/hdulist.py

```diff
@@ -322,13 +322,14 @@ def __getitem__(self, key):
 
     def __contains__(self, item):
         """
-        Returns `True` if ``HDUList.index_of(item)`` succeeds.
+        Returns `True` if ``HDUList.index_of(item)`` or
+        ``HDUList.index(item)`` succeeds.
         """
 
         try:
             self._try_while_unread_hdus(self.index_of, item)
         except KeyError:
-            return False
+            return super().__contains__(item)
```

Comment:
I think it would make sense to do this directly in `self.index_of`, which could also take an HDU instance as parameter,  as it then does a loop on loaded HDUs. This would also probably avoid to load all HDUs when it is not needed.

Reply:
@saimn I thought of that too but I decided against this for two reasons: `index__of(key)` describes what `key` is and allowing e.g. `ImageHDU` as a key seems weird. Also, `list.index()` actually does this. Also, I am not sure what `self._try_while_unread_hdus()` does but it sounds to me like this is not going to solve the problem of not loading HDUs.

Reply:
`_try_while_unread_hdus` repeats the function given as first argument while loading one more HDU at each iteration, as long as there are more HDUs to load. 

> Also, list.index() actually does this.

True indeed, but having this feature in `index_of` would allow to stop loading HDUs when the searched one is found (when using `self._try_while_unread_hdus(self.index_of, item)`).

Reply:
It is not difficult to implement this in the `index_of()` - I'll be happy to do it. But the thing that in order to do something like `hdulist[0] in hdulist`, `hdulist[0]` must already be loaded. However, it makes no sense to load any HDU if it is not already loaded - **only when** checking for inclusion of HDU objects (by `id` - not by `ext` name/ver). What I am trying to say is: 1. assume `hdulist = fits.open('somefile'); hdu = fits.ImageHDU();` 2. Then `hdu in hdulist` should always fail because there is no way for me to create an `hdu` that will have the signature of the HDU that will be loaded from file *in the future*. 3. The inclusion test *may pass* only if i) I create `hdu` as `hdu = hdulist[key]` OR ii) I create HDU list from a list of *existing* HDUs: `hdulist = HDUList([hdu, ...])`.

Reply:
@saimn That is, this could be implemented in `index_of()` [which, IMO should have been just `index()`] but also in `__contains__()` then there should be a check like this: `if isinstance(item, _BaseHDU): return self.index_of(item) else: self._try_while_unread_hdus(self.index_of, item)`. This will avoid loading HDUs when item is an HDU.

Reply:
> ...having this feature in index_of would allow to stop loading HDUs when the searched one is found...

That is, what I am saying is that HDUs should not be loaded at all if they have not been already loaded when `item` is an HDU.

Reply:
If `item` is an HDU, loaded from the HDUList on which you are calling `__contains__`, then `self._try_while_unread_hdus(self.index_of, item)` will find it directly without needing to load more HDUs (they are loaded sequentially, one by one).
So I think the easiest would be to add an instance identity check in the for loop in `index_of`, and maybe rename the `key` argument to `key_or_hdu` ?